src/lib/data/alignment.js

```diff
@@ -331,12 +371,80 @@ export default class Alignment {
    * @returns {Boolean} true - groups were merged, false - was not
    */
   mergeActiveGroupWithAnotherByToken (token) {
-    if (this.activeAlignmentGroup && this.tokenIsGrouped(token)) {
+    if (this.hasActiveAlignment && this.tokenIsGrouped(token)) {
       const tokensGroup = this.findAlignmentGroup(token)
+
+      const indexDeleted = this.removeGroupFromAlignmentGroups(tokensGroup)
       this.removeGroupFromAlignmentIds(tokensGroup)
-      this.activeAlignmentGroup.merge(tokensGroup)
+
+      this.activeAlignmentGroup.merge(tokensGroup, indexDeleted)
       return true
     }
     return false
   }
+
+  /**
+   * Step back inside active group
+   * If we step back merge step, then we would insert unmerged group back to the list
+   */
+  undoInActiveGroup () {
+    if (!this.hasActiveAlignment) {
+      return
+    }
+
+    const results = this.activeAlignmentGroup.undo()
+    if (results && results.length > 0) {
+      for (let i = 0; i < results.length; i++) {
+        this.insertUnmergedGroup(results[i])
+      }
+    }
+  }
+
+  /**
+   *
+   * @param {Object} data
+   *        {AlignmentGroup} data.tokensGroup
+   *        {Number} data.indexDeleted
+   */
+  insertUnmergedGroup (data) {
+    this.alignmentGroups.splice(data.indexDeleted, 0, data.tokensGroup)

```

Comment:
It works this way:
- we have `[ group1, group2, group 3 ]`
- we merge `group1` inside `group 2` - merge step would return `{ tokensGroup: group1, indexDeleted: 0 }`; list - `[ group2, group3 ]`
- we merge `group3` inside `group2` - merge step would return `{ tokensGroup: group3, indexDeleted: 1 }`; list - `[ group2 ]`
- we unmerge `group3` - would `insertUnmergedGroup` `group3` on the `index: 1`; list - `[ group2, group3 ]`
- we unmerge `group1` - would `insertUnmergedGroup` `group1` on the `index: 0`; list - `[ group1, group2, group3 ]`

This way we could return to the initial groups order to save the steps through groups

Reply:
ah, I missed that you were not actually removing any elements. Is there any chance that groups could be added or removed before the redo and the indexDeleted would no longer be the right place to insert the unmerged group?

Reply:
oh, it is an interesting question.
I think that it could be broken in theory only if we continue to create new groups while we have undone groups.
This could mix the order.
I believe that we should do here similiar to steps - we truncate steps if we start to add new tokens to the group.
Also we should truncate undone groups we we start to create the new one.
As for previosly merge/unmerge groups - it shouldn't break anything - because we don't have any variants to add group not to the end of the list.

Reply:
ok.